src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/ExtensibleSourceRepositoryProvider.cs

```diff
@@ -36,20 +40,22 @@ public ExtensibleSourceRepositoryProvider()
         /// </summary>
         [ImportingConstructor]
         public ExtensibleSourceRepositoryProvider([ImportMany] IEnumerable<Lazy<INuGetResourceProvider>> resourceProviders, [Import] Configuration.ISettings settings)
-            : this(new Configuration.PackageSourceProvider(settings, migratePackageSources: null), resourceProviders)
         {
+            _settings = settings;
+            _resourceProviders = Repository.Provider.GetVisualStudio().Concat(resourceProviders);
         }
 
-        /// <summary>
-        /// Non-MEF constructor
-        /// </summary>
-        public ExtensibleSourceRepositoryProvider(Configuration.IPackageSourceProvider packageSourceProvider, IEnumerable<Lazy<INuGetResourceProvider>> resourceProviders)
+        private void EnsureInitialized()
         {
-            _packageSourceProvider = packageSourceProvider;
-            _resourceProviders = Repository.Provider.GetVisualStudio().Concat(resourceProviders);
+            if (!_initialized)
+            {
+                _initialized = true;
+
+                _packageSourceProvider = new Configuration.PackageSourceProvider(_settings, migratePackageSources: null);
```

Comment:
You can declare `packageSourceProvider` as Lazy and use this code block as predicate. It will be thread-safe. And no extra `EnsureInitialize` method will be needed.